components/homme/src/theta-l/share/imex_mod.F90

```diff
@@ -310,33 +319,36 @@ subroutine compute_stage_value_dirk(nm1,alphadt_nm1,n0,alphadt_n0,np1,dt2,elem,h
 
                    do k = 1,nlev-1
                       dphi(i,j,k) = dphi_n0(i,j,k) + &
-                           dt2*g*((w_np1(i,j,k+1) - w_np1(i,j,k)) + &
-                           alpha*(x(i,j,k+1) - x(i,j,k)))
+                           dt2*g_from_phi((elem(ie)%state%phinh_i(i,j,k,nt)+&
+                         elem(ie)%state%phinh_i(i,j,k+1,nt))/2.0)*((w_np1(i,j,k+1) - w_np1(i,j,k)) + alpha*(x(i,j,k+1) - x(i,j,k))) !DA_CHANGE
                    end do
-                   dphi(i,j,nlev) = dphi_n0(i,j,nlev) - dt2*g*(w_np1(i,j,nlev) + alpha*x(i,j,nlev))
+                   dphi(i,j,nlev) = dphi_n0(i,j,nlev) - dt2*g_from_phi((elem(ie)%state%phinh_i(i,j,k,nt)+  elem(ie)%state%phinh_i(i,j,k+1,nt))/2.0)*(w_np1(i,j,nlev) + alpha*x(i,j,nlev)) !DA_CHANGE
                 end do
              end do
           end if
           do k = 1,nlev
              w_np1(:,:,k) = w_np1(:,:,k) + alphas*x(:,:,k)
           end do
-
+          do k=nlev,1,-1  ! scan
+             phi_np1(:,:,k) = phi_np1(:,:,k+1)-dphi(:,:,k)
+          enddo
+ 
           call pnh_and_exner_from_eos2(hvcoord,elem(ie)%state%vtheta_dp(:,:,:,np1),&
-               elem(ie)%state%dp3d(:,:,:,np1),dphi,pnh,exner,dpnh_dp_i,'dirk2')
-          Fn(:,:,1:nlev) = w_np1(:,:,1:nlev) - (w_n0(:,:,1:nlev) + g*dt2 * (dpnh_dp_i(:,:,1:nlev)-1))
+               elem(ie)%state%dp3d(:,:,:,np1),dphi,phi_np1,pnh,exner,dpnh_dp_i,'dirk2')
+          Fn(:,:,1:nlev) = w_np1(:,:,1:nlev) - (w_n0(:,:,1:nlev) + g_from_phi((elem(ie)%state%phinh_i(:,:,1:nlev,nt)+ &
+                         elem(ie)%state%phinh_i(:,:,2:nlevp,nt))/2.0, nlev)*dt2 * (dpnh_dp_i(:,:,1:nlev)-1)) ! DA_CHANGE
 
           ! this is not used in this loop, so move out of loop
           !reserr=maxval(abs(Fn))/(wmax*abs(dt2)) 
           deltaerr=maxval(abs(x))/wmax
-
           ! update iteration count and error measure
           itercount=itercount+1
           !if (reserr < restol) exit
           if (deltaerr<deltatol) exit
        end do ! end do for the do while loop
 
        ! update phi:
```

Comment:
minor note -- imex is already quite expensive and using extra functions in it will slow it down